src/precice/impl/DataContext.cpp

```diff
@@ -2,80 +2,103 @@
 #include <memory>
 #include "mesh/Data.hpp"
 #include "mesh/Mesh.hpp"
+#include "time/Waveform.hpp"
 
 namespace precice {
 namespace impl {
 
 DataContext::DataContext(mesh::PtrData data, mesh::PtrMesh mesh)
 {
+  PRECICE_TRACE();
   PRECICE_ASSERT(data);
+  time::PtrWaveform ptrWaveform(new time::Waveform(data->values().size(), DataContext::EXTRAPOLATION_ORDER, DataContext::INTERPOLATION_ORDER));
```

Comment:
Not sure where we should create and store the waveform(s). The current implementation looks to me redundant. We currently have four different places, where a waveform is created (see below). Read and Write mapping are exclusive, so this is fine, but depending on the setup the waveform stored in the `BaseCouplingScheme` and one of the waveforms stored in the data context might be representing the same data.

I think it might be a good idea to move ownership and creating of all waveforms to the `DataContext` and provide the pointer to the `BaseCouplingScheme`, if required.

### Create the `DataContext::_providedWaveform`

https://github.com/BenjaminRodenberg/precice/blob/5ce6dcc149a9ccccefe5aa5a205654b19b197a8a/src/precice/impl/DataContext.cpp#L14-L15

### Create the `DataContext::_fromWaveform`, if this participant uses read mapping for data

https://github.com/BenjaminRodenberg/precice/blob/5ce6dcc149a9ccccefe5aa5a205654b19b197a8a/src/precice/impl/DataContext.cpp#L128-L129

### Create the `DataContext::_toWaveform`, if this participant uses write mapping

https://github.com/BenjaminRodenberg/precice/blob/5ce6dcc149a9ccccefe5aa5a205654b19b197a8a/src/precice/impl/DataContext.cpp#L139-L140

### Add to `BaseCouplingScheme::_waveforms` for `_allData` data

https://github.com/BenjaminRodenberg/precice/blob/5ce6dcc149a9ccccefe5aa5a205654b19b197a8a/src/cplscheme/BaseCouplingScheme.cpp#L458-L460

Reply:
Goal: Only call `new time::Waveform(...)` in `TimeConfiguration/WaveformConfiguration` and hand over pointer to modules using `Waveform`. We might need a `Waveform::initialize(int size)`, because `data->values().size() ` is not known at configuration time.

Reply:
I will continue working on this problem here: https://github.com/precice/precice/pull/1133

The final outcome should be merged back into this PR.